New field type: LSB Extended Field

tigre-bleu opened the pull request:
**Description:**

This PR adds a new type of field that I add to implement to dissect a custom protocol. It is not very often seen in protocols but as it is generic, maybe it worth including in the main scapy distribution.

Up to you to decide.

**Details:**

This type of field has a variable number of bytes. Size is not taken from another field. Each byte is defined as follows:
- The 7 MSB bits are data
- The LSB is an extension bit
  - 0 means it is last byte of the field ("stopping bit")
  - 1 means there is another byte after this one ("forwarding bit")

To get the actual data, it is necessary to hop the binary data from byte to byte until LSB is 0.

**Tests:**

I tested with success both on Python 2 and Python 3. I did not write unit tests as to be fair I don't really know how to do it. However, I am not sure that it is very important as the field does not rely on anything special from Scapy.

*Note:*
*I noticed while writing this PR description that my VSCode automatically removed the trailing spaces from the modified files. It is always better for the code itself but it clutters a little the diff. Sorry for that.*

gpotter2 commented:
Not sure yet if we should include that. Do you have any examples of protocols in which it is used ? If we do want to include this, it would be great if it could also support MSB

Also you would need to make it work on Python 3 (I see `"".join()` that won’t work on Py3) and add tests (wait that we all agree for this to be included before working on it :-))

tigre-bleu commented:
> Not sure yet if we should include that. Do you have any examples of protocols in which it is used ? If we do want to include this, it would be great if it could also support MSB

I implemented this field to study the [All-purpose structured EUROCONTROL surveillance information exchange (ASTERIX)](https://www.eurocontrol.int/services/asterix) protocol. It is used for exchanges between radars and air traffic control/management centers.

I am not sure that it is implemented elsewhere but I can create a MSBExtendedField should that help anybody.

> Also you would need to make it work on Python 3 (I see `"".join()` that won’t work on Py3) and add tests (wait that we all agree for this to be included before working on it :-))

I am not sure what you meant about the `join()` function. It works with my Python 3.7. `"".join()` function [still exists](https://docs.python.org/3/library/stdtypes.html#str.join). Maybe it was a confusion with the deprecated `string.join()`?

For the tests, I will wait & see :-)

gpotter2 commented:
> I am not sure what you meant about the join() function. It works with my Python 3.7. "".join() function still exists. Maybe it was a confusion with the deprecated string.join()?

Mhmh confused huh ? :smile: You would need bytes thus `b"".join(...)`

I’m gonna go ahead and approve this, it seems it’s a quite common field after all. It would be great if you could also add MSB (maybe share some functions ?), and add tests.

Scapy’s tests use a custom historical format, which is pretty self-explicit. You could add your tests to `scapy/test/fields.uts`

Thanks again for your contribution !

PS: you’ll notice we have unit tests running on. We also enforce PEP8 & a codespell. Here are the PEP8 results of the PR in its current state:
```
scapy/fields.py:2304:26: E226 missing whitespace around arithmetic operator
scapy/fields.py:2342:17: E226 missing whitespace around arithmetic operator
```

tigre-bleu commented:
Ok, I modified the field to a generic `BitExtendedField(extension_bit)` so that we can set the FX bit on any 8 bit of the byte.

I created `LSBExtendeField` and `MSBExtendedField` as 2 subclasses of `BitExtendedField`.

I will write some tests, push them and ping you when it's done so you can review.

tigre-bleu commented:
I merged my branch with upstream scapy and added the tests to `fields.uts.`

I added some validation on the field:  `m2i()` returns `None` when the field is malformed (ie no Stopping Bit). I felt it was better than returning a wrong value, but please correct me if it was a wrong assumption.

Everything looks fine for me on Python 2 and Python 3 and all tests passed, so I think this PR is ready for review.

Thanks

gpotter2 commented:
@tigre-bleu Great PR thanks ! Could you please fix the confclits ?
@guedou Do we allow doc spacing changes ?

guedou commented:
@gpotter2 as a rule of thumb: no. Useless modifications made the review harder.

@tigre-bleu could you remove the useless spacing changes, and squash you commits? Thanks.

tigre-bleu commented:
@guedou It's done.
